pkg/mgmt/volume/volume.go

```diff
@@ -49,31 +51,116 @@ func (c *VolController) syncHandler(key string) error {
+	err = runtimenew.DefaultUnstructuredConverter.FromUnstructured(unstructuredVol.UnstructuredContent(), &vol)
+	if err != nil {
+		klog.Infof("err %s, While converting unstructured obj to typed object\n", err.Error())
+	}
+	VolCopy := vol.DeepCopy()
 	err = c.syncVol(VolCopy)
 	return err
 }
 
+// addVol is the add event handler for LVMVolume
+func (c *VolController) addVol(obj interface{}) {
+	Vol, ok := c.getStructuredObject(obj)
+	if !ok {
+		runtime.HandleError(fmt.Errorf("Couldn't get Vol object %#v", obj))
+		return
+	}
+
+	if lvm.NodeID != Vol.Spec.OwnerNodeID {
+		return
+	}
+	klog.Infof("Got add event for Vol %s", Vol.Name)
+	klog.Infof("lvmvolume object to be enqueued by Add handler: %v", Vol)
+	c.enqueueVol(Vol)
+}
+
+// updateVol is the update event handler for LVMVolume
+func (c *VolController) updateVol(oldObj, newObj interface{}) {
+	newVol, ok := c.getStructuredObject(newObj)
+	if !ok {
+		runtime.HandleError(fmt.Errorf("Couldn't get Vol object %#v", newVol))
+		return
+	}
+
+	if lvm.NodeID != newVol.Spec.OwnerNodeID {
+		return
+	}
+
+	if c.isDeletionCandidate(newVol) {
+		klog.Infof("Got update event for deleted Vol %s, Deletion timestamp %s", newVol.Name, newVol.ObjectMeta.DeletionTimestamp)
+		c.enqueueVol(newVol)
+	}
+}
+
+// deleteVol is the delete event handler for LVMVolume
+func (c *VolController) deleteVol(obj interface{}) {
+	Vol, ok := c.getStructuredObject(obj)
+	if !ok {
+		unstructuredObj, ok := obj.(*unstructured.Unstructured)
+		if ok {
+			tombStone := cache.DeletedFinalStateUnknown{}
+			err := runtimenew.DefaultUnstructuredConverter.FromUnstructured(unstructuredObj.UnstructuredContent(), &tombStone)
+			if err != nil {
+				runtime.HandleError(fmt.Errorf("couldn't get object from tombstone %#v", obj))
+				return
+			}
+			Vol, ok = tombStone.Obj.(*apis.LVMVolume)
+			if !ok {
+				runtime.HandleError(fmt.Errorf("tombstone contained object that is not a lvmvolume %#v", obj))
+				return
+			}
+		} else {
+			runtime.HandleError(fmt.Errorf("couldnt type assert obj: %#v to unstructured obj", obj))
+			return
+		}
+	}
+
+	if lvm.NodeID != Vol.Spec.OwnerNodeID {
+		return
+	}
+
+	klog.Infof("Got delete event for Vol %s", Vol.Name)
+	c.enqueueVol(Vol)
+}
+
 // enqueueVol takes a LVMVolume resource and converts it into a namespace/name
 // string which is then put onto the work queue. This method should *not* be
 // passed resources of any type other than LVMVolume.
 func (c *VolController) enqueueVol(obj interface{}) {
-	var key string
-	var err error
-	if key, err = cache.MetaNamespaceKeyFunc(obj); err != nil {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
 		runtime.HandleError(err)
 		return
 	}
 	c.workqueue.Add(key)
+}
 
+//Obj from queue is not readily in lvmvol type. This function would convert obj into lvmvolume type.
+func (c *VolController) getStructuredObject(obj interface{}) (*apis.LVMVolume, bool) {
+	unstructuredInterface, ok := obj.(*unstructured.Unstructured)
+	if ok {
```

Comment:
Default case should not be error handling.
Better to use...
```go
if !ok {
        // Handle error
}

// Code for when type assertion succeeds
```

Reply:
Not sure what do you mean. We are just checking if we are able to assert object to unstructured type. If yes, then using standard function to convert that into required resource type.